File: 474-lowest-common-ancestor-ii/standard_answer.cc
/**
 * Definition of ParentTreeNode:
 * class ParentTreeNode {
 * public:
 *     int val;
 *     ParentTreeNode *parent, *left, *right;
 * }
 */


class Solution {
public:
    /*
     * @param root: The root of the tree
     * @param A: node in the tree
     * @param B: node in the tree
     * @return: The lowest common ancestor of A and B
     */
    ParentTreeNode * lowestCommonAncestorII(ParentTreeNode * root, ParentTreeNode * A, ParentTreeNode * B) {
        // write your code here
        vector<ParentTreeNode*> APath = nodeToRoot(A);
        vector<ParentTreeNode*> BPath = nodeToRoot(B);
        int indexA = APath.size() - 1;
        int indexB = BPath.size() - 1;
        ParentTreeNode* lowestCommonAncestor = NULL;
        while (indexA >= 0 && indexB >= 0) {
            if (APath[indexA] != BPath[indexB]) {
                break;
            }
            lowestCommonAncestor = APath[indexA];
            indexA--;
            indexB--;
        }
        return lowestCommonAncestor;
    }
    vector<ParentTreeNode*> nodeToRoot(ParentTreeNode* node) {
        vector<ParentTreeNode*> result;
        while (node != NULL) {
            result.push_back(node);
            node = node->parent;
        }
        return result;
    }
};

File: 645-find-the-celebrity/standard_answer.java
/* The knows API is defined in the parent class Relation.
      boolean knows(int a, int b); */

public class Solution extends Relation {
    /**
     * @param n a party with n people
     * @return the celebrity's label or -1
     */
    public int findCelebrity(int n) {
        // Write your code here
        int s = 0;
        for (int i = 1; i < n; i++) {
            if (knows(s, i)) {
                s = i;
            }
        }
        for (int i = 0; i < n; i++) {
            if (s == i) {
                continue;
            }
            if (knows(s, i)) {
                return -1;
            }
            if (!knows(i, s)) {
                return -1; 
            }
        }
        return s;
    }
}
